broadinstitute/single_cell_portal_core
# Complete integration for author-generated differential expression results (SCP-5096, SCP-5005)

#### BACKGROUND & CHANGES
This update completes the integration work for author-generated differential expression data.  Now, any study owner can upload their own results and have then viewable upon successful ingest.  This work handles the normal ingest submission/polling and creation of `DifferentialExpressionResult` entries upon success.  Further refactoring work has been done in `DifferentialExpressionResult` to reduce redundancy/complexity in various areas (mostly setting output file names/paths).  This update also handles the normal cleanup of results (including manifests), as well as deleting any auto-calculated DE results that may have existed already on the study.

Note: When viewing results in the UI, the scores will all load as `NaN` as the table view updates are handled in https://github.com/broadinstitute/single_cell_portal_core/pull/1813.  Additionally, there is an existing bug with this file and `author_de.py` where the pairwise results for `eosinophils` and `CSN1S1 macrophages` cannot be viewed as the sorting of the labels is incorrect.  It will not show up in the UI, and when deleting results the file will persist in the bucket as we enforce natural sorting on all labels.

### MANUAL TESTING
1. Download the test file from https://raw.githubusercontent.com/broadinstitute/scp-ingest-pipeline/development/tests/data/author_de/author_de_test_data_human_milk_All_Cells_UMAP_General_celltype.csv
2. Boot as normal and load the upload wizard for `Human milk - differential expression` study
3. Add the above file as a DE file, selecting `All Cells UMAP` and `General_Celltype` as the cluster & annotation, respectively
4. Upload and confirm in `development.log` that the file launches an ingest job:
```
Request object sent to Google Life Sciences API, excluding 'environment' parameters:
---
:actions:
  :commands:
  - python
  - ingest_pipeline.py
  - "--study-id"
  - 62718fbb94ec8f51542049d6
  - "--study-file-id"
  - 649c5dd33d68b745848569b4
  - "--user-metrics-uuid"
  - 0fc4d8e5-6c4c-422e-94a3-c78801757202
  - ingest_differential_expression
  - "--annotation-name"
  - General_Celltype
  - "--annotation-scope"
  - study
  - "--annotation-type"
  - group
  - "--cluster-name"
  - All Cells UMAP
  - "--differential-expression-file"
  - gs://fc-32035468-639c-414f-8275-6d3998cfe355/author_de_test_data_human_milk_All_Cells_UMAP_General_celltype.csv
  - "--study-accession"
  - SCP66
  - "--method"
  - wilcoxon
  - "--ingest-differential-expression"
  :image_uri: gcr.io/broad-singlecellportal-staging/scp-ingest-pipeline:1.27.1
  :labels: {}
  :timeout: {}
...
```
5. Open the GCP bucket for this study (link is in the `Study details` page) and confirm you still see old result files still in `_scp_internal/differential_expression`
6. Wait for the ingest to complete and confirm you get a success email:
```
Your Single Cell Portal parse job has completed with the following results:

Total parse time: 1 Minute and 5 Seconds
Differential expression ingest completed for General_Celltype
One-vs-rest comparisons: B cells, CSN1S1 macrophages, dendritic cells
Total pairwise comparisons: 3
```
7. Load the explore tab and confirm that you can load one-vs-rest results for `B cells`, `CSN1S1 macrophages`, and `dendritic cells`
8. Confirm that `B cells` has pairwise results for  `CSN1S1 macrophages` and `dendritic cells`
9. Go back to the bucket and confirm there are now only the following files:
![Screenshot 2023-06-28 at 11 15 02 AM](https://github.com/broadinstitute/single_cell_portal_core/assets/729968/a7bc12dd-6b04-47d3-990e-59f4fc8e78c1)
10. Optional: go back to the upload wizard and delete the DE file
11. Confirm in `development.log` that you see files being removed:
```
Removing DE output SCP66:General_Celltype--group--study at _scp_internal/differential_expression/All_Cells_UMAP--General_Celltype--B_cells--study--wilcoxon.tsv
Removing DE output SCP66:General_Celltype--group--study at _scp_internal/differential_expression/All_Cells_UMAP--General_Celltype--CSN1S1_macrophages--study--wilcoxon.tsv
Removing DE output SCP66:General_Celltype--group--study at _scp_internal/differential_expression/All_Cells_UMAP--General_Celltype--dendritic_cells--study--wilcoxon.tsv
Removing DE output SCP66:General_Celltype--group--study at _scp_internal/differential_expression/All_Cells_UMAP--General_Celltype--B_cells--CSN1S1_macrophages--study--wilcoxon.tsv
Removing DE output SCP66:General_Celltype--group--study at _scp_internal/differential_expression/All_Cells_UMAP--General_Celltype--B_cells--dendritic_cells--study--wilcoxon.tsv
Removing manifest for SCP66:General_Celltype--group--study at _scp_internal/differential_expression/All_Cells_UMAP--General_Celltype--manifest.tsv
```
12. Refresh the GCP bucket and confirm all files were deleted (except for the `eosinophils-CSN1S1 macrophages` pairwise result
